### DISCUSSION TITLE
How to delete a file from storage server-side?

### DISCUSSION THREAD

**Discussion by christianfl** [supabase/supabase#8061] [Questions] [answered]
Hello there! I was curious about this storage handling I observed:

We have a trigger which after uploading of a new avatar image deletes all other avatars from the same user in our avatar bucket:

```
create or replace function private.delete_old_avatars()
returns trigger
language plpgsql security definer
as
$$
begin
  delete from storage.objects
  where owner = new.owner and bucket_id = 'avatars' and id <> new.id;
  return new;
end;
$$;

create trigger on_avatar_uploaded
  after insert on storage.objects
  for each row
  when (new.bucket_id = 'avatars')
  execute procedure private.delete_old_avatars();
```

The rows in the storage table get deleted and on hosted supabase I am not able to view the old images anymore.

**However, on self hosted, I observed that the files itself are still there in the file system. Is this intended? How could I delete those?**

Thanks in advance,
Christian

**Comment by GaryAustin1** [comment, marked as answer]
Deleting the object from objects table does not remove the file from s3 storage.
You need to actually use the storage API to delete the files.

I delete them using the http extension when a row in another table containing the image is deleted.
This has the code:  https://github.com/supabase/supabase/discussions/7067 that should be close to what you need if you want to do it from within the database.
Otherwise you will need to handle it in an external cloud or edge function.

**Reply by christianfl** [reply]
Ah! Did not see the relating items. Thank you very much!

**Comment by ghost** [comment]
@GaryAustin1 Is this still the case ?
Since the storage can be used through the S3 API, I've done some testing and from what I've seen, deleting from SQL with ‘delete from storage.objects where name = “xxx”’ deletes it completely.

**Reply by GaryAustin1** [reply]
I've not not seen anything on this changing.  The s3 interface still goes thru storage.objects because of security/ownership test.
